### PR TITLE
SQL: Allow intervals to be passed as request parameters

### LINKED ISSUE #45915: SQL: interval types not accepted as parameters in ODBC
**Elasticsearch version** (`bin/elasticsearch --version`): 8.0.0-SNAPSHOT
**Description of the problem including expected versus actual behavior**:

Providing an interval as a parameter type yields exception.

**Steps to reproduce**:
```
POST /_sql
{
  "query": "SELECT null WHERE 0 = ?",
  "params": [{"type":"interval_year", "value":"P326Y"}]
}
```

**Provide logs (if relevant)**:

```
{
  "error": {
    "root_cause": [
      {
        "type": "sql_illegal_argument_exception",
        "reason": "cannot convert from [keyword] to [interval_year]"
      }
    ],
    "type": "parsing_exception",
    "reason": "line 1:24: Unexpected actual parameter type [KEYWORD] for type [interval_year]",
    "caused_by": {
      "type": "sql_illegal_argument_exception",
      "reason": "cannot convert from [keyword] to [interval_year]"
    }
  },
  "status": 400
}
```

### PR DESCRIPTION
**Pull request by bpintea**
This PR adds the possibility for intervals to be passed as request
parameters. This functionality is only strictly required by the drivers.

The PR also changes the way the protocol tests are executed,
allowing to check not only that values of the SQL type are correctly
returned, but also that parameters of those types are correctly received
by the server.

Closes #45915

### CODE REVIEW

File: x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/SqlProtocolTestCase.java [hunk 1]
```diff
@@ -244,22 +263,21 @@ private void assertQuery(String sql, String columnName, String columnType, Objec
             options.addHeader("Accept", randomFrom("*/*", "application/" + format));
             request.setOptions(options);
         }
-        if ((false == columnar && randomBoolean()) || columnar) {
+        if (columnar || randomBoolean()) {
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/SqlProtocolTestCase.java]: @astefan, Do you remember the purpose of the previous status of this if?

**Review by matriv — APPROVED** [review]
LGTM. Left a comment/question.

File: x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/SqlProtocolTestCase.java (hunk 1 above)
Comment (astefan) [reply x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/SqlProtocolTestCase.java]: @matriv, yes. The condition says "set the columnar parameter to true, if columnar value is true; set the columnar parameter to false **explicitly** if randomBoolean is true or do nothing (columnar remains at the default value - false) if randomBoolean is false". And I believe @bpintea 's simplification is correct.

**Review by astefan — APPROVED** [review]
LGTM

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java [hunk 2]
```diff
@@ -721,13 +721,25 @@ public Literal visitParamLiteral(ParamLiteralContext ctx) {
         }
         // otherwise we need to make sure that xcontent-serialized value is converted to the correct type
         try {
-            if (canConvert(sourceType, dataType) == false) {
+            Object literalValue;
+            // The interval types are handled "directly", as they must not be exposed through the SQL types converter, which would then in
+            // turn expose the "protocol" interval types (i.e. "interval_year" & co.) to, for instance, SQL scalars (like CAST) a.s.o.
+            if (SqlDataTypes.isYearMonthInterval(dataType)) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java]: I think this logic could be encapsulated into `Intervals` class:

```java
if (SqlDataTypes.isInterval(dataType)) {
    return Intervals.from(param.value, dataType);
}
```

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/literal/interval/IntervalYearMonth.java [hunk 3]
```diff
@@ -34,6 +34,15 @@ public IntervalYearMonth(StreamInput in) throws IOException {
         super(period(in), SqlDataTypes.fromTypeName(in.readString()));
     }
 
+    public static IntervalYearMonth from(String value, DataType intervalType) {
+        return new IntervalYearMonth(Period.parse(value), intervalType);
+    }
+
+    @Override
+    public int compareTo(Interval<Period> o) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/literal/interval/IntervalYearMonth.java]: Why is this needed?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/literal/interval/Interval.java [hunk 4]
```diff
@@ -26,7 +26,8 @@
  * Unfortunately because the SQL interval type is not preserved accurately by the JDK TemporalAmount class
  * in both cases, the data type needs to be carried around as it cannot be inferred.
  */
-public abstract class Interval<I extends TemporalAmount> implements ConstantNamedWriteable, ToXContentObject, IntervalScripting {
+public abstract class Interval<I extends TemporalAmount> implements ConstantNamedWriteable, ToXContentObject, IntervalScripting,
+        Comparable<Interval<I>> {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/literal/interval/Interval.java]: Why an interval has to be comparable?

**Review by costin — COMMENTED** [review]
Left some comments

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/literal/interval/Interval.java (hunk 4 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/literal/interval/Interval.java]: In particular, this is part of the change applied to the `SqlProtocolTestCase`, which performs an equality test between the value in the SELECT and same value (plus its corresponding type), but passed as a parameter; which in turns verifies that accepting parameters of the tested types works.

In general, this also makes comparisons between intervals possible. (Other DBs, like Postgres, also implement this, so it shouldn't be that far fetched.) Currently a simple SELECT with an intervals comparison returns `null` (b/c that's what the `Comparisons.compare` returns if types are neither `Comparable`, nor `Number`; but this might be a different topic).

As far as the Java backing types go, it's a split case:
- `Duration` also implements `Comparable`;
- `Period` thought, doesn't, but that's because daylight savings are taking into account with the day component; however, this only backs up YEAR-MONTH intervals - i.e. no day component - so the extension in the patch is safe for these intervals.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java (hunk 2 above)
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java]: This would look better there indeed! Pushed an update, hope it matches your suggestion.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/literal/interval/Interval.java (hunk 4 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/literal/interval/Interval.java]: I would fix it in a separate PR to address the comparison between intervals and add proper testing.
Even more so considering `Period` doesn't implement `Comparable`.
In testing we only use the equality not comparison check . 

If comparing Period proves to be tricky, I propose handling this by preventing the comparison to be done.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/literal/interval/Intervals.java [hunk 5]
```diff
@@ -97,6 +97,16 @@ public static TemporalAmount of(Source source, long duration, TimeUnit unit) {
         }
     }
 
+    public static Interval<?> from(String isoValue, DataType dataType) {
+        if (isYearMonthInterval(dataType)) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/literal/interval/Intervals.java]: I would move the `from` methods from the classes here as well. Mainly because those classes have behavior and a contract (same methods, generics) while this utility class only static methods.
Also doing a type check to then do `TypeIChecked.doSomething` defeats the whole purpose of putting said method in the source class as it's not generic.
Further more the methods are one liners..

**Comment by matriv** [comment]
@elastic/es-ql

**Comment by bpintea** [comment]
Will break this PR into subsequent ones.
